lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,208 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of BitCast to nonstandard types. LLVM
+// transformations bitcast some vector types to scalar types, which are not
+// universally supported across all targets. We need ensure that "optimized"
+// LLVM IR doesn't have primitive types other than supported by the
+// SPIR target (i.e. "scalar 8/16/32/64-bit integer and 16/32/64-bit floating
+// point types, 2/3/4/8/16-element vector of scalar types").
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spv-lower-bitcast-to-nonstandard-type"
+
+#include "SPIRVInternal.h"
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+#include <stack>
+#include <utility>
+
+using namespace llvm;
+
+namespace {
+
+class SPIRVLowerBitCastToNonStandardTypeBase {
+public:
+  SPIRVLowerBitCastToNonStandardTypeBase() {}
+
+  bool runLowerBitCastToNonStandardType(Module &Module) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known
+    auto *M = &Module;
+    bool Changed = false;
+
+    std::vector<std::pair<Instruction *, VectorType *>> BCastsToNonStdVec;
+    std::stack<Instruction *> InstsToErase;
+    for (auto &F : M->functions())
+      for (auto &BB : F)
+        for (auto &I : BB)
+          if (BitCastInst *BC = dyn_cast<BitCastInst>(&I)) {
+            auto *DestTy = BC->getDestTy();
+            if (auto *ElemTy = dyn_cast<PointerType>(DestTy))
+              DestTy = ElemTy->getElementType();
+            if (auto *DestVecTy = dyn_cast<VectorType>(DestTy)) {
+              uint64_t NumElemsInDestVec =
+                  DestVecTy->getElementCount().getValue();
+              if (NumElemsInDestVec != 2 && NumElemsInDestVec != 3 &&
+                  NumElemsInDestVec != 4 && NumElemsInDestVec != 8 &&
+                  NumElemsInDestVec != 16)
+                BCastsToNonStdVec.push_back(std::make_pair(&I, DestVecTy));
+            }
+          }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    while (!InstsToErase.empty()) {
+      InstsToErase.top()->eraseFromParent();
+      InstsToErase.pop();
+    }
+
+    verifyRegularizationPass(*M, "SPIRVLowerBitCastToNonStandardType");
+    return Changed;
+  }
+
+  bool lowerBitCastToNonStdVec(std::pair<Instruction *, VectorType *> &Inst,
+                               std::stack<Instruction *> &InstsToErase) {
+    bool Changed = false;
+    std::vector<Instruction *> InstsToInsert;
+    Instruction *I = Inst.first;
+    VectorType *DestVecTy = Inst.second;
+    uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+    InstsToErase.push(I);
+    AddrSpaceCastInst *Src{nullptr};
```

Comment:
Other variables seem to be initialized via old fashioned assignment style, why so is here?
```suggestion
    AddrSpaceCastInst *Src = nullptr;
```